File: modules/graphics/src/main/java/com/wiranoid/fuzzy/graphics/Window.java
package com.wiranoid.fuzzy.graphics;

import com.wiranoid.fuzzy.core.utils.Disposable;
import org.lwjgl.glfw.*;
import org.lwjgl.opengl.GL;

import static org.lwjgl.glfw.GLFW.*;
import static org.lwjgl.opengl.GL11.GL_FALSE;
import static org.lwjgl.system.MemoryUtil.NULL;

public class Window implements Disposable {
    private final long id;

    private GLFWKeyCallback keyCallback;
    private GLFWCursorPosCallback mouseCallback;
    private GLFWScrollCallback scrollCallback;

    private boolean vsync;
    private String title;

    public long getId() {
        return id;
    }

    public void setKeyCallback(GLFWKeyCallback keyCallback) {
        this.keyCallback = keyCallback;
        glfwSetKeyCallback(id, keyCallback);
    }

    public void setMouseCallback(GLFWCursorPosCallback mouseCallback) {
        this.mouseCallback = mouseCallback;
        glfwSetCursorPosCallback(id, mouseCallback);
    }

    public void setScrollCallback(GLFWScrollCallback scrollCallback) {
        this.scrollCallback = scrollCallback;
        glfwSetScrollCallback(id, scrollCallback);
    }

    public boolean isVSyncEnabled() {
        return vsync;
    }

    public void setVSync(boolean vsync) {
        this.vsync = vsync;
        if (vsync) {
            glfwSwapInterval(1);
        } else {
            glfwSwapInterval(0);
        }
    }

    public String getTitle() {
        return title;
    }

    public void setTitle(String title) {
        glfwSetWindowTitle(id, title);
    }

    public Window(int width, int height, String title, boolean isFullScreen, boolean vsync) {
        this.vsync = vsync;
        this.title = title;

        glfwDefaultWindowHints();
        glfwWindowHint(GLFW_CONTEXT_VERSION_MAJOR, 3);
        glfwWindowHint(GLFW_CONTEXT_VERSION_MINOR, 3);
        glfwWindowHint(GLFW_OPENGL_PROFILE, GLFW_OPENGL_CORE_PROFILE);
        glfwWindowHint(GLFW_RESIZABLE, GL_FALSE);

        if (isFullScreen) {
            id = glfwCreateWindow(width, height, title, glfwGetPrimaryMonitor(), NULL);
        } else {
            id = glfwCreateWindow(width, height, title, NULL, NULL);

            /* Center window on screen */
            GLFWVidMode vidMode = glfwGetVideoMode(glfwGetPrimaryMonitor());
            glfwSetWindowPos(id,
                    (vidMode.width() - width) / 2,
                    (vidMode.height() - height) / 2);
        }

        if (id == NULL) {
            glfwTerminate();
            throw new RuntimeException("Failed to create GLFW window");
        }

        /* Create OpenGL context */
        glfwMakeContextCurrent(id);
        GL.createCapabilities();

        /* Enable v-sync */
        if (vsync) {
            glfwSwapInterval(1);
        }
    }

    public void setInputMode(int mode, int value) {
        glfwSetInputMode(id, mode, value);
    }

    public boolean isClosing() {
        return glfwWindowShouldClose(id);
    }

    public void update() {
        glfwSwapBuffers(id);
    }

    public void dispose() {
        glfwDestroyWindow(id);
        keyCallback.free();
        mouseCallback.free();
        scrollCallback.free();
    }

}

File: modules/graphics/src/Window.cpp
#include "../include/Window.h"
#include <stdexcept>
#include <string>

using namespace graphics;

Window::Window(int width,
               int height,
               std::string title,
               bool isFullScreen,
               bool vsync) :
        _width(width), _height(height), _title(title), _vsync(vsync) {
    if (isFullScreen) {
        this->_window = glfwCreateWindow(width, height, title.c_str(), glfwGetPrimaryMonitor(), nullptr);
    } else {
        this->_window = glfwCreateWindow(width, height, title.c_str(), nullptr, nullptr);

        // Center window on screen
        auto vidMode = glfwGetVideoMode(glfwGetPrimaryMonitor());
        glfwSetWindowPos(this->_window,
                         (vidMode->width - width) / 2,
                         (vidMode->height - height) / 2);
    }

    if (this->_window == nullptr) {
        throw std::runtime_error("Failed to create GLFW window");
    }

    // Create OpenGL context
    glfwMakeContextCurrent(this->_window);

    // Enable v-sync
    if (vsync) {
        glfwSwapInterval(1);
    }

    // Define the viewport dimensions
    int frameBufferWidth, frameBufferHeight;
    glfwGetFramebufferSize(this->_window, &frameBufferWidth, &frameBufferHeight);
    glViewport(0, 0, frameBufferWidth, frameBufferHeight);
}

void Window::setKeyCallback(GLFWkeyfun callback) {
    glfwSetKeyCallback(this->_window, callback);
}

void Window::setMouseCallback(GLFWcursorposfun callback) {
    glfwSetCursorPosCallback(this->_window, callback);
}

void Window::setScrollCallback(GLFWscrollfun callback) {
    glfwSetScrollCallback(this->_window, callback);
}

bool Window::isClosing() {
    return (bool) glfwWindowShouldClose(this->_window);
}

void Window::setIsShoudClose(bool isShouldClose) {
    glfwSetWindowShouldClose(this->_window, isShouldClose);
}

bool Window::isVSyncEnabled() {
    return this->_vsync;
}

void Window::setVSync(bool vsync) {
    this->_vsync = vsync;
    if (vsync) {
        glfwSwapInterval(1);
    } else {
        glfwSwapInterval(0);
    }
}

void Window::makeContextCurrent() {
    glfwMakeContextCurrent(this->_window);
}

void Window::setInputMode(int mode, int value) {
    glfwSetInputMode(this->_window, mode, value);
}

void Window::swapBuffers() {
    glfwSwapBuffers(this->_window);
}

void Window::destroy() {
    glfwDestroyWindow(this->_window);
}

Window::~Window() {
    //empty destructor
}
